### DISCUSSION TITLE
Issues during LoRA Fine-Tuning: 'MatmulLtState' object has no attribute 'memory_efficient_backward'

### DISCUSSION THREAD

**Discussion by anika-n-sharma** [community/community#148200] [Programming Help]
### Body

I am experimenting with LoRA to fine-tune a model to process and analyze PDF files so that I can ask questions based on the files. Essentially, I would upload PDFs, then the program would split it into chunks, and "learn" from the PDFs so I wouldn't have to repeatedly upload files and it would remember the context from the files (as I am building a streamlit application) and then, generate a vector store for querying.

`def fine_tune_model_lora_with_suggestions(train_data):
    st.write("Starting high-performance fine-tuning with LoRA...")
    try:
        import torch
        from transformers import AutoModelForCausalLM, AutoTokenizer, TrainingArguments, Trainer
        from peft import LoraConfig, get_peft_model
        from datasets import Dataset

        # Define model name
        model_name = "bigscience/bloom-7b1"

        # Load tokenizer
        tokenizer = AutoTokenizer.from_pretrained(model_name)

        # Load model normally if CUDA is available, otherwise use CPU
        if torch.cuda.is_available():
            model = AutoModelForCausalLM.from_pretrained(
                model_name,
                load_in_8bit=True,  # Enable 8-bit quantization
                device_map="auto",  # Automatically map layers between GPU and CPU
                llm_int8_enable_fp32_cpu_offload=True,  # Offload some layers to CPU in FP32
                torch_dtype=torch.float16,  # Use FP16 for GPU-loaded layers
            )
        else:
            model = AutoModelForCausalLM.from_pretrained(model_name)

        # Apply LoRA configuration
        lora_config = LoraConfig(
            r=16,
            lora_alpha=32,
            lora_dropout=0.05,
            bias="none",
            task_type="CAUSAL_LM",
            target_modules=["query_key_value"],  # Specify target modules for LoRA
        )
        model = get_peft_model(model, lora_config)

        # Prepare dataset
        dataset = Dataset.from_list(train_data)

        def tokenize_function(examples):
            tokens = tokenizer(
                examples["text"],
                padding="max_length",
                truncation=True,
                max_length=512,
            )
            tokens["labels"] = tokens["input_ids"].copy()
            return tokens

        tokenized_dataset = dataset.map(tokenize_function, batched=True)

        # Training arguments
        training_args = TrainingArguments(
            per_device_train_batch_size=4,
            gradient_accumulation_steps=4,
            max_steps=200,
            learning_rate=2e-4,
            fp16=torch.cuda.is_available(),  # Enable FP16 only if CUDA is available
            logging_steps=10,
            output_dir="./outputs",
            save_steps=10,
            save_total_limit=2,
            report_to="none",
        )

        # Initialize Trainer
        trainer = Trainer(
            model=model,
            args=training_args,
            train_dataset=tokenized_dataset,
        )

        # Train model
        trainer.train()

        # Save fine-tuned model
        model.save_pretrained("./fine_tuned_bloom_lora")
        tokenizer.save_pretrained("./fine_tuned_bloom_lora")
        st.write("Fine-tuning completed successfully.")

    except ImportError as e:
        st.error(f"Import Error: {e}")
    except Exception as e:
        st.error(f"Error during LoRA fine-tuning: {e}")`

Also, here are the libraries I have installed: 

`!pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118

!pip install transformers==4.45.2 sentence-transformers==3.1.1

!pip install bitsandbytes --extra-index-url https://huggingface.github.io/bitsandbytes/
!pip install peft==0.2.0 datasets==2.7.1 einops==0.6.1

!pip install PyPDF2 langchain langchain_community streamlit faiss-cpu python-dotenv google-generativeai pyngrok wandb
!pip install jedi fsspec[http]==2024.9.0

!pip install git+https://github.com/huggingface/accelerate.git

!sudo apt-get install -y libcairo2-dev libjpeg-dev libpng-dev libfreetype6-dev
!pip install pycairo

!pip install appdirs fire tensorboardX gradio

`

I would appreciate any help I could get on this! Thank you!

I initially thought that the error is occuring due to calling old versions of libraries or if certain functions were depreciated, but I am not so sure. I have tried pip install bitsandbytes == 0.37.2, but it still gives the same errors.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by jakubbbdev2** [comment]
Hi,

I noticed you're having an issue with the MatmulLtState error during LoRA fine-tuning. Here are a few things you can try:

PEFT Version: Try installing an earlier version of peft:

pip install peft==0.1.0
Update Libraries: Make sure transformers and torch are up-to-date:

pip install --upgrade transformers torch
Memory Efficient Backward: The error could be due to a missing feature in your setup. Ensure your model and libraries support it.

Dependencies: Reinstall everything in a clean virtual environment to avoid conflicts.

CUDA Setup: Double-check that your GPU setup is correct if you're using CUDA.

Try these steps and let me know how it goes!

**Comment by github-actions** [comment]
**💬 Your Product Feedback Has Been Submitted 🎉**

Thank you for taking the time to share your insights with us! Your feedback is invaluable as we build a better GitHub experience for all our users.

**Here's what you can expect moving forward ⏩**
- Your input will be carefully reviewed and cataloged by members of our product teams. 
  - Due to the high volume of submissions, we may not always be able to provide individual responses.
  - Rest assured, your feedback will help chart our course for product improvements.
- Other users may engage with your post, sharing their own perspectives or experiences. 
- GitHub staff may reach out for further clarification or insight. 
  - We may 'Answer' your discussion if there is a current solution, workaround, or roadmap/changelog post related to the feedback.

**Where to look to see what's shipping 👀**
- Read the [Changelog](https://github.blog/changelog/) for real-time updates on the latest GitHub features, enhancements, and calls for feedback.
- Explore our [Product Roadmap](https://github.com/orgs/github/projects/4247), which details upcoming major releases and initiatives.

**What you can do in the meantime 💻**
- Upvote and comment on other user feedback Discussions that resonate with you.
- Add more information at any point! Useful details include: use cases, relevant labels, desired outcomes, and any accompanying screenshots.

As a member of the GitHub community, your participation is essential. While we can't promise that every suggestion will be implemented, we want to emphasize that your feedback is instrumental in guiding our decisions and priorities.

Thank you once again for your contribution to making GitHub even better! We're grateful for your ongoing support and collaboration in shaping the future of our platform. ⭐

**Comment by mecodeatlas** [comment]
Hey @anika-n-sharma! 👋 
We're going through our unanswered discussions, I know it has been a minute, but I wanted to check and see if you were able to get this solved?

**Comment by mecodeatlas** [comment]
Hi @anika-n-sharma! We hope you found what you were looking for. ✨

We're closing this discussion because it has not had recent activity and/or other members have asked for more information to assist you but received no response. Thank you for helping us maintain a productive and tidy community for all our members.
